### PR TITLE
`TextControl`: Add opt-in prop for 40px default size

### PR DESCRIPTION
**Pull request by brookewp**
Part of https://github.com/WordPress/gutenberg/issues/46741

## What?

Add a new opt-in prop `__next40pxDefaultSize` to TextControl, following the plan outlined in above mentioned PR. 

## Why?

For more consistency in styling. 

## How?

A new, temporary `__next40pxDefaultSize` prop. When the prop is set to `true`, the input will have a height of 40px instead of 29px. 

### In Storybook: 

1. `npm run storybook:dev`
2. By default, TextControl will have a height of 29px and `__next40pxDefaultSize` set to `false`
3. Set `__next40pxDefaultSize` to `true`
4. The input height of the component should now be 40px

### In the editor

Smoke test the component in the editor; TextControl shouldn't have any visible changes for now.

### CODE REVIEW

File: packages/components/src/text-control/stories/index.story.tsx
Comment (ciampo) [inline packages/components/src/text-control/stories/index.story.tsx]: Since we're using TypeScript, this change shouldn't be needed

File: packages/components/src/text-control/index.tsx
Comment (ciampo) [inline packages/components/src/text-control/index.tsx]: It looks like `classname` is already passed to `BaseControl`, so probably we shouldn't pass it to the `input` element too?

File: packages/editor/src/components/post-publish-panel/test/__snapshots__/index.js.snap
Comment (ciampo) [inline packages/editor/src/components/post-publish-panel/test/__snapshots__/index.js.snap]: This snapshot having changed is also an indication that the classnames change above is introducing some side effects

**Review by ciampo — COMMENTED** [review]
Left a couple of comments, but this is looking good!

cc'ing @jameskoster and @richtabor

**Comment by jameskoster** [comment]
Nice. 

Should we also update that 29px default height to 32px here (related: https://github.com/WordPress/gutenberg/pull/55490)?

**Review by chad1008 — COMMENTED** [review]
Looking good to me! No new comments to add, but I'll let @ciampo approve after previous feedback is addressed.

**Comment by brookewp** [comment]
>Should we also update that 29px default height to 32px here (related: https://github.com/WordPress/gutenberg/pull/55490)?

From the PR you mentioned:

>@ciampo how do you think we should proceed here? I don't think an increase of 2px warrants a temporary prop or other affordance due to the minimal impact.

With the above context, I assume a 3px change is also okay to proceed with?

**Comment by ciampo** [comment]
> With the above context, I assume a 3px change is also okay to proceed with?

Hey @jameskoster , to make sure we're on the same page here: in this PR we are working towards transitioning `TextControl`'s default size to `40px`. We can definitely update the current default size to `32px`, but when the transition is over, all instances of `TextControl` will render with a height of `40px` (meaning no `32px` size). Is that ok for you?

**Comment by jameskoster** [comment]
Hm, thinking beyond the Editor, forcing 40px in _all_ situations might be a bit heavy-handed? It's fine as a default, but there could be scenarios that call for a smaller control. cc @jasmussen @SaxonF @richtabor for more thoughts on that.

**Comment by jasmussen** [comment]
The idea is 40px is the default, but the compact 32px can be used on a per-basis case. 40px is the tall and graceful, neatly clickable on big screens and phones, default. I'd turn the question around, why would you want a 32px default?

**Comment by jameskoster** [comment]
> meaning no 32px size

Unless I'm mistaken I think the question was about whether 32px is needed at all, not whether it's the default.

**Comment by jasmussen** [comment]
32 feels useful across buttons and text controls both, as they might appear next to each others, IMO.

**Comment by ciampo** [comment]
> 32 feels useful across buttons and text controls both, as they might appear next to each others, IMO.

This sounds sensible. However, since I don't remember this being included in [the initial specs](https://github.com/WordPress/gutenberg/issues/46734), it may be beneficial to understand what other components we may need to render at a `32px` size (apart from `Button`).

We can keep this PR as-is (focused on starting the transition to `40px` default size), and later make any necessary changes for a potential `32px` variant.

**Comment by jameskoster** [comment]
I'd imagine any control that we're making 40px could also use a 32px size variant.

**Comment by ciampo** [comment]
> I'd imagine any control that we're making 40px could also use a 32px size variant.

Even more so, then, let's discuss this holistically so that we can have as much of a coherent approach as possible, since this feels like a departure from the specs shared [here](https://github.com/WordPress/gutenberg/issues/46734), where `32px` felt like an expectation only for button and toggles.

For this PR, I think it's fine to update the current default size from 29 to 32px, knowing that it will be eventually become 40px.

**Review by chad1008 — APPROVED** [review]
This LGTM :rocket: :ship:

**Pull request merged**
